src/chat.rs

```diff
@@ -0,0 +1,217 @@
+use serde::{Deserialize, Serialize};
+use std::time::Duration;
+
+use crate::error::AnyError;
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct Usage {
+    pub prompt_tokens: i64,
+    pub completion_tokens: i64,
+    pub total_tokens: i64,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct MessageResponse {
+    pub role: String,
+    pub content: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct Choice {
+    pub message: MessageResponse,
+    pub finish_reason: String,
+    pub index: i64,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct SuccessResponse {
+    pub id: Option<String>,
+    pub object: String,
+    pub created: i64,
+    pub model: String,
+    pub usage: Usage,
+    pub choices: Vec<Choice>,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct Error {
+    pub message: String,
+    pub r#type: String,
+    pub param: String,
+    pub code: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct ErrorResponse {
+    pub error: Error,
+}
+
+impl From<ErrorResponse> for String {
+    fn from(value: ErrorResponse) -> Self {
+        format!(
+            "Error: {} {} {} {}",
+            value.error.message, value.error.r#type, value.error.param, value.error.code
+        )
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+#[serde(untagged)]
+pub enum Response {
+    Success(SuccessResponse),
+    Error(ErrorResponse),
+}
+
+impl TryFrom<&str> for Response {
+    type Error = serde_json::Error;
+
+    fn try_from(value: &str) -> Result<Self, serde_json::Error> {
+        serde_json::from_str(value)
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub enum Model {
+    #[serde(rename = "gpt-3.5-turbo")]
+    Gpt3Turbo,
+    #[serde(rename = "gpt-4")]
+    Gpt4,
+}
+
+impl Model {
+    pub fn new(model: &str) -> Result<Self, AnyError> {
+        match model {
+            "gpt-3.5-turbo" => Ok(Self::Gpt3Turbo),
+            "gpt-4" => Ok(Self::Gpt4),
+            _ => Err("Error: Not a valid model name".into()),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct MessageRequest {
```

Comment:
This type is the same as the response one. Are they broken out to support a future where they might differ?